File: iris-windows/src/services/model-price-comparison.ts
/**
 * The side-by-side under "How Iris answers" — the Windows half of
 * `ModelPriceComparison.swift`.
 *
 * NO PRICE LIVES IN THIS APP. Every number comes from
 * `GET {publik}/api/iris/model-prices` (publik's `lib/iris-model-prices.ts`),
 * which reads publik's own tier prices from the code that bills them and
 * copies Anthropic's and OpenAI's from their published pages with the date
 * they were read. When the route cannot be reached the settings window says so
 * and shows no number at all — a stale price shown as current is exactly the
 * dishonest thing this exists to avoid.
 *
 * One Windows-specific step, `comparisonForThisPc`: this client always asks
 * publik API for `publik-balanced`, and its own-key picker offers different
 * models from the Mac's. So the comparison shown here is the balanced level,
 * with the own-key row repriced for the model the reader actually picked —
 * from the same response's `anthropicListPrices` and its stated assumption,
 * never from a number compiled in.
 */

export interface ComparisonAssumption {
  questionsPerMonth: number;
  inputTokensPerQuestion: number;
  outputTokensPerQuestion: number;
  summary: string;
}

export interface ComparisonOption {
  provider: "publik-api" | "anthropic-key" | "codex";
  label: string;
  model: string;
  servedBy: string | null;
  billing: "per_token" | "subscription";
  inputUsdPerMillion: number | null;
  outputUsdPerMillion: number | null;
  estimatedCostPerQuestionUsd: number | null;
  estimatedMonthlyUsd: number;
  note: string;
  source: string;
}

export interface ComparisonLevel {
  tier: "fast" | "balanced" | "smart";
  options: ComparisonOption[];
}

export interface AnthropicListPrice {
  model: string;
  inputUsdPerMillion: number;
  cacheWriteUsdPerMillion: number;
  cacheReadUsdPerMillion: number;
  outputUsdPerMillion: number;
}

export interface ModelPriceComparison {
  version: 1;
  assumption: ComparisonAssumption;
  caveat: string;
  levels: ComparisonLevel[];
  anthropicListPrices: AnthropicListPrice[];
}

function isFiniteNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function isNullableNumber(value: unknown): value is number | null {
  return value === null || isFiniteNumber(value);
}

function isOption(value: unknown): value is ComparisonOption {
  if (!value || typeof value !== "object") return false;
  const option = value as Record<string, unknown>;
  return (
    (option.provider === "publik-api" || option.provider === "anthropic-key" || option.provider === "codex") &&
    typeof option.label === "string" &&
    typeof option.model === "string" &&
    (option.servedBy === null || typeof option.servedBy === "string") &&
    (option.billing === "per_token" || option.billing === "subscription") &&
    isNullableNumber(option.inputUsdPerMillion) &&
    isNullableNumber(option.outputUsdPerMillion) &&
    isNullableNumber(option.estimatedCostPerQuestionUsd) &&
    isFiniteNumber(option.estimatedMonthlyUsd) &&
    typeof option.note === "string" &&
    typeof option.source === "string"
  );
}

/**
 * The route's JSON, checked field by field. Anything unexpected is `null` —
 * the settings window then says prices are unavailable rather than rendering
 * half a comparison.
 */
export function parseModelPriceComparison(body: unknown): ModelPriceComparison | null {
  if (!body || typeof body !== "object") return null;
  const candidate = body as Record<string, unknown>;
  if (candidate.version !== 1) return null;

  const assumption = candidate.assumption as Record<string, unknown> | undefined;
  if (
    !assumption ||
    !isFiniteNumber(assumption.questionsPerMonth) ||
    !isFiniteNumber(assumption.inputTokensPerQuestion) ||
    !isFiniteNumber(assumption.outputTokensPerQuestion) ||
    typeof assumption.summary !== "string"
  ) {
    return null;
  }
  if (typeof candidate.caveat !== "string") return null;

  if (!Array.isArray(candidate.levels) || candidate.levels.length === 0) return null;
  const levels: ComparisonLevel[] = [];
  for (const level of candidate.levels as unknown[]) {
    const record = level as Record<string, unknown> | null;
    if (!record || (record.tier !== "fast" && record.tier !== "balanced" && record.tier !== "smart")) return null;
    if (!Array.isArray(record.options) || !record.options.every(isOption)) return null;
    levels.push({ tier: record.tier, options: record.options as ComparisonOption[] });
  }

  const anthropicListPrices: AnthropicListPrice[] = [];
  if (Array.isArray(candidate.anthropicListPrices)) {
    for (const row of candidate.anthropicListPrices as unknown[]) {
      const price = row as Record<string, unknown> | null;
      if (
        price &&
        typeof price.model === "string" &&
        isFiniteNumber(price.inputUsdPerMillion) &&
        isFiniteNumber(price.cacheWriteUsdPerMillion) &&
        isFiniteNumber(price.cacheReadUsdPerMillion) &&
        isFiniteNumber(price.outputUsdPerMillion)
      ) {
        anthropicListPrices.push(price as unknown as AnthropicListPrice);
      }
    }
  }

  return {
    version: 1,
    assumption: {
      questionsPerMonth: assumption.questionsPerMonth,
      inputTokensPerQuestion: assumption.inputTokensPerQuestion,
      outputTokensPerQuestion: assumption.outputTokensPerQuestion,
      summary: assumption.summary,
    },
    caveat: candidate.caveat,
    levels,
    anthropicListPrices,
  };
}

/**
 * The list price for a model id as the picker stores it, dated snapshot
 * included: "claude-opus-4-1-20250805" is "claude-opus-4-1". Longest match
 * wins, so "claude-opus-4-1" never resolves to a shorter family row.
 */
export function listPriceForModel(comparison: ModelPriceComparison, modelId: string): AnthropicListPrice | null {
  const matches = comparison.anthropicListPrices.filter(
    (row) => modelId === row.model || modelId.startsWith(`${row.model}-`)
  );
  if (matches.length === 0) return null;
  return matches.reduce((longest, row) => (row.model.length > longest.model.length ? row : longest));
}

function roundTo(value: number, places: number): number {
  const factor = 10 ** places;
  return Math.round(value * factor) / factor;
}

/**
 * The comparison as this PC sees it: the balanced level (the tier this client
 * always sends publik API), with the own-key row for the reader's own model.
 * Null when the response has no balanced level. When the reader's model has
 * no listed price, the own-key row is dropped rather than shown with the
 * wrong model's numbers.
 */
export function comparisonForThisPc(
  comparison: ModelPriceComparison,
  readersAnthropicModel: string
): { assumption: ComparisonAssumption; caveat: string; options: ComparisonOption[] } | null {
  const balanced = comparison.levels.find((level) => level.tier === "balanced");
  if (!balanced) return null;

  const options = balanced.options.flatMap((option): ComparisonOption[] => {
    if (option.provider !== "anthropic-key") return [option];
    const price = listPriceForModel(comparison, readersAnthropicModel);
    if (!price) return [];
    const { inputTokensPerQuestion, outputTokensPerQuestion, questionsPerMonth } = comparison.assumption;
    const perQuestion =
      (inputTokensPerQuestion * price.inputUsdPerMillion + outputTokensPerQuestion * price.outputUsdPerMillion) / 1_000_000;
    return [
      {
        ...option,
        model: price.model,
        inputUsdPerMillion: price.inputUsdPerMillion,
        outputUsdPerMillion: price.outputUsdPerMillion,
        estimatedCostPerQuestionUsd: roundTo(perQuestion, 4),
        estimatedMonthlyUsd: roundTo(perQuestion * questionsPerMonth, 2),
      },
    ];
  });

  return { assumption: comparison.assumption, caveat: comparison.caveat, options };
}

File: iris-macos/leanring-buddy/ModelPriceComparison.swift
//
//  ModelPriceComparison.swift
//  leanring-buddy
//
//  The side-by-side Iris shows next to "How Iris answers": publik API against
//  the reader's own Anthropic key against a ChatGPT plan (the Codex route), in
//  dollars per million tokens and an estimated dollars a month.
//
//  NO PRICE LIVES IN THIS APP. Every number comes from
//  `GET {publik}/api/iris/model-prices` (publik's `lib/iris-model-prices.ts`),
//  which reads publik's own tier prices from the code that bills them and
//  copies Anthropic's and OpenAI's from their published pages with the date
//  they were read. When that route cannot be reached the view says so; it
//  never falls back to a number compiled into the binary, because a stale
//  price shown as current is exactly the dishonest thing this exists to avoid.
//
//  The same response carries Anthropic's list prices per model, and the spend
//  ledger prices the reader's own-key calls from them once they have loaded
//  (`AssistantServerPublishedPrices`) — so "what your key spent" and "what
//  your key would cost" come from the same rows.
//

import Combine
import Foundation

/// The route's JSON, decoded as-is.
nonisolated struct ModelPriceComparison: Decodable, Equatable, Sendable {

    struct Assumption: Decodable, Equatable, Sendable {
        let questionsPerMonth: Int
        let inputTokensPerQuestion: Int
        let outputTokensPerQuestion: Int
        let summary: String
    }

    struct Option: Decodable, Equatable, Sendable, Identifiable {
        /// "publik-api", "anthropic-key" or "codex" — `AssistantProviderPreference`'s raw values.
        let provider: String
        let label: String
        let model: String
        /// The model that actually answers, when it differs from `model`.
        let servedBy: String?
        /// "per_token" or "subscription".
        let billing: String
        let inputUsdPerMillion: Double?
        let outputUsdPerMillion: Double?
        let estimatedCostPerQuestionUsd: Double?
        let estimatedMonthlyUsd: Double
        let note: String
        let source: String

        var id: String { provider }
        var isBilledPerToken: Bool { billing == "per_token" }
    }

    struct Level: Decodable, Equatable, Sendable {
        /// "fast", "balanced" or "smart".
        let tier: String
        let options: [Option]

        func option(for provider: UsageProvider) -> Option? {
            options.first { $0.provider == provider.rawValue }
        }
    }

    struct AnthropicListPrice: Decodable, Equatable, Sendable {
        let model: String
        let inputUsdPerMillion: Double
        let cacheWriteUsdPerMillion: Double
        let cacheReadUsdPerMillion: Double
        let outputUsdPerMillion: Double
    }

    let version: Int
    let assumption: Assumption
    let caveat: String
    let levels: [Level]
    let anthropicListPrices: [AnthropicListPrice]

    /// The level matching the model the reader picked in Iris, by the same
    /// mapping the gateway alias uses (`UsageModelTier.forIrisModelName`).
    func level(forIrisModelName irisModelName: String) -> Level? {
        let tier = UsageModelTier.forIrisModelName(irisModelName)
        return levels.first { $0.tier == tier.rawValue }
    }

    /// The published rows as the spend ledger's pricing type. A price is
    /// carried through its decimal text so $3.75 stays $3.75 rather than
    /// the nearest binary fraction.
    var anthropicPricingByModel: [String: AssistantModelPricing] {
        var byModel: [String: AssistantModelPricing] = [:]
        for row in anthropicListPrices {
            guard let input = Self.exactDecimal(row.inputUsdPerMillion),
                  let cacheWrite = Self.exactDecimal(row.cacheWriteUsdPerMillion),
                  let cacheRead = Self.exactDecimal(row.cacheReadUsdPerMillion),
                  let output = Self.exactDecimal(row.outputUsdPerMillion) else { continue }
            byModel[row.model] = AssistantModelPricing(
                inputPerMillion: input,
                cacheWritePerMillion: cacheWrite,
                cacheReadPerMillion: cacheRead,
                outputPerMillion: output
            )
        }
        return byModel
    }

    static func exactDecimal(_ value: Double) -> Decimal? {
        guard value.isFinite, value >= 0 else { return nil }
        return Decimal(string: String(value))
    }
}

/// Fetches the comparison once per launch, in the background. Nothing awaits
/// it: the panel renders whatever state it is in.
@MainActor
final class ModelPriceComparisonStore: ObservableObject {

    enum LoadState: Equatable {
        case notLoaded
        case loading
        case loaded(ModelPriceComparison)
        /// The route could not be reached or answered something unreadable.
        case unavailable
    }

    @Published private(set) var loadState: LoadState = .notLoaded

    var comparison: ModelPriceComparison? {
        if case .loaded(let comparison) = loadState { return comparison }
        return nil
    }

    private let endpoint: URL
    private let fetchData: @Sendable (URL) async throws -> Data

    init(
        publikBaseURL: URL,
        fetchData: @escaping @Sendable (URL) async throws -> Data = ModelPriceComparisonStore.fetchOverTheNetwork
    ) {
        self.endpoint = publikBaseURL.appendingPathComponent("api/iris/model-prices")
        self.fetchData = fetchData
    }

    /// Starts a fetch unless one has succeeded or is running. A failure can be
    /// retried by calling this again (the panel does, when it opens).
    func loadIfNeeded() {
        switch loadState {
        case .loaded, .loading:
            return
        case .notLoaded, .unavailable:
            break
        }
        loadState = .loading
        let endpoint = self.endpoint
        let fetchData = self.fetchData
        Task { [weak self] in
            let decoded: ModelPriceComparison?
            do {
                let data = try await fetchData(endpoint)
                decoded = try JSONDecoder().decode(ModelPriceComparison.self, from: data)
            } catch {
                decoded = nil
            }
            guard let self else { return }
            if let decoded, decoded.version == 1, !decoded.levels.isEmpty {
                self.loadState = .loaded(decoded)
                AssistantServerPublishedPrices.install(decoded.anthropicPricingByModel)
            } else {
                self.loadState = .unavailable
            }
        }
    }

    nonisolated static func fetchOverTheNetwork(_ url: URL) async throws -> Data {
        var request = URLRequest(url: url)
        request.timeoutInterval = 10
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, (200..<300).contains(httpResponse.statusCode) else {
            throw URLError(.badServerResponse)
        }
        return data
    }
}
